cmd/heartbeat/heartbeat.go

```diff
@@ -282,3 +262,32 @@ func setLogFields(params paramscmd.Params) {
 		log.WithField("is_write", params.Heartbeat.IsWrite)
 	}
 }
+
+func handleHeartbeatError(v *viper.Viper, queueFilepath string, err error) (int, error) {
+	var errauth api.ErrAuth
+
+	// api.ErrAuth represents an error when parsing api key.
+	// Save heartbeats to offline db even when api key invalid.
+	// It avoids losing heartbeats when api key is invalid.
+	if errors.As(err, &errauth) {
+		if err := offlinecmd.SaveHeartbeats(v, nil, queueFilepath); err != nil {
+			log.Errorf("failed to save heartbeats to offline queue: %s", err)
+		}
+
+		return errauth.ExitCode(), fmt.Errorf("sending heartbeat(s) failed: %s", errauth.Message())
+	}
+
+	if errbackoff, ok := err.(api.ErrBackoff); ok {
+		log.Debugf("sending heartbeat(s) failed: %s", errbackoff.Message())
+		return errbackoff.ExitCode(), nil
+	}
+
+	if errwaka, ok := err.(wakaerror.Error); ok {
+		return errwaka.ExitCode(), fmt.Errorf("sending heartbeat(s) failed: %s", errwaka.Message())
+	}
```

Comment:
It's not needed to filter for `api.ErrBackoff` since it implements wakaerror.Error interface. I also wouldn't move to a separate function as it moves few lines only and the `Run` method is already a small one, around 35 lines.

Reply:
Linter fails when it's back in the original function.

I'm catching backoff separately here to suppress the error log by returning `nil` instead of `error`.